nextcord/application_command.py

```diff
@@ -743,16 +743,39 @@ def from_callback(
             #  might be able to do something here.
             if option_class:
                 skip_counter = 1
-                # TODO: use typing.get_type_hints when 3.9 is standard
-                typehints = typing_extensions.get_type_hints(self.callback, include_extras=True)
                 # Getting the callback with `self_skip = inspect.ismethod(self.callback)` was problematic due to the
                 #  decorator going into effect before the class is instantiated, thus being a function at the time.
                 #  Try to look into fixing that in the future?
                 #  If self.parent_cog isn't reliable enough, we can possibly check if the first parameter name is `self`
                 if self.parent_cog:
                     skip_counter += 1
 
-                for name, param in signature(self.callback).parameters.items():
+                # TODO: use typing.get_type_hints when 3.9 is standard
+                typehints = typing_extensions.get_type_hints(self.callback, include_extras=True)
+                callback_params = signature(self.callback).parameters
+                non_option_params = sum(
+                    # could be a self or interaction parameter
+                    param.annotation is not param.empty
+                    # will always be an interaction parameter
+                    and not issubclass(param.annotation, Interaction)
+                    # will always be a self parameter
+                    # TODO: use typing.Self when 3.11 is standard
+                    and param.annotation is not typing_extensions.Self
+                    # will always be a self parameter
+                    and not isinstance(param.annotation, typing.TypeVar)
+                    for param in list(callback_params.values())[:skip_counter]
+                )
```

Comment:
Sorry for changing it so much, but when testing it I noticed it failing on some cases and it seems like it actually should be negated again (remove the `not`s and change `and` back to `or`).

That way it will count how many parameters *could* be self or interaction.

```suggestion
                non_option_params = sum(
                    # could be a self or interaction parameter
                    param.annotation is param.empty
                    # will always be an interaction parameter
                    or issubclass(param.annotation, Interaction)
                    # will always be a self parameter
                    # TODO: use typing.Self when 3.11 is standard
                    or param.annotation is typing_extensions.Self
                    # will always be a self parameter
                    or isinstance(param.annotation, typing.TypeVar)
                    for param in list(callback_params.values())[:skip_counter]
                )
```